### PR TITLE
dlgUseDate changes

### LINKED ISSUE #4603: Continue improvements in the use date dialogue
1. The addition of the quarter function has been a good improvement.
2. The main suggestions I make now are to improve the layout of the check-boxes.
3. In addition, I think the Pentad does not work correctly.  It should produce a factor with 72 levels, and with just 5 (or 6) days in each month.  Perhaps Beth can look quickly at this?
4. And more of the variables should have an option of labels.  For example Quarter is just 1 2, 3, 4 which is correct if numeric.  It could be Jan-Mar, Apr-Jun, Jul-Sep, Oct-Dec as short labels.  And these should (like the months), ideally be done so that they are automatically in the computer's language where it is not English.  
5. Similarly Dekad would have the label option of Jan1, Jan2, Jan3, Feb1... With Pentad it would be Jan1 to Jan6. 
6. And more variables should be shifted as an option, namely Month, Pentad and Dekad.  When these are labelled they will look the same, but Aug would be level 1, if shifted to Aug, with July as level 12.  For Dekads Aug1 would become Level 1, with Jul1 being level 34.

So my suggested layout is roughly as shown by Danny who may take a  photo.  The 6 variables that can be shifted are first, each with a column, namely 
**Year, Month, Day in Year(366), Pentad, Dekad, Quarter.**
In the left hand side are the 3 options **Numeric, Abb Name, Name**, with check-boxes to indicate what is possible.
So:
a) Year is just numeric,
b) Month is Numeric, or Abb Name or Name.
c) Day in Year(366) is just numeric
d) Pentad, Dekad, Quarter are Numeric or Abb Name.

Below that block is:
**Shifted Month and pull-down** as we have.

Below that is another block of variables to which shifted does not apply, i.e. 
a) Day in Month - just numeric
b) Day in Year - just numeric
c) Week is numeric or abbrev name Wk1, Wk52 or full name Week1, to Week52.
d) Weekday also has the 3 options
e) Leap Year is just numeric (it is actually logical, but let's not be too purist!

I suggest that Patrick could work on the layout and organisation of the dialogue, while Beth concentrates (with him) on the R code.

### PR DESCRIPTION
**Pull request by AlexSananka**
@africanmathsinitiative/developers this is ready for review. It should fix #4603

### CODE REVIEW

**Comment by rdstern** [comment]
The change of the month to a factor is working fine.  It is fine with January as the first month and also with the shifted year..
So is the pentad and dekad factors, when January if the first month.

If the year is shifted the dekads and pentads no longer work.  They give missing values.  They were ok (I think) as character, but perhaps that was never tested._

**Comment by AlexSananka** [comment]
@rdstern this should now be fine. Please have a look.

**Comment by rdstern** [comment]
I have checked and it is working well now.  I can give you 2 options now.  The first is that this is ready for @dannyparsons .  The second is another (small?) addition.
Only even think of doing this if you agree it is small and quick!  It is to add a third option for each of dekad and pentad.  It is the same as you have, but the factor labels only go from 1 to 3 (dekads) each month, or 1 to 6 for pentads.  So it would go Jan1, Jan2, Jan3, Feb1, Feb2, Feb3, etc, rather than the current Feb4, Feb5 and Feb6.
They would be another checkbox in the "Name" row.  If you do this, then I would change Name to Alt. (for alternative).
If you get the numeric value from the one you already have, then it is just (x-1)%% 3 +1 for dekads or (x-1)%%6 +1 for pentads.  But may be simpler to get the index directly.

**Comment by dannyparsons** [comment]
I think the current headers for the rows are useful for month, year etc. so I'm hesitant to change them.

What we could do easily is make the other option - Jan1, Jan2, Jan3, Feb1, Feb2, Feb3 etc. the default.
@AlexSananka you just need to modify the numbers as Roger showed above. The levels then also need to be adjusted to have a repeated 1:3 instead of 1:36.

**Comment by AlexSananka** [comment]
@dannyparsons I have implemented your suggested options. @rdstern please test that this is OK.

**Comment by rdstern** [comment]
Great. Let's merge this!

File: instat/static/InstatObject/R/data_object_R6.R [hunk 1]
```diff
@@ -2162,7 +2162,7 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
     self$append_to_variables_metadata(col_names = col_name, property = doy_start_label, new_val = s_start_day)
   }
   if(month_abbr) {
-    month_abbr_vector <- forcats::fct_shift(f = lubridate::month(col_data, label = TRUE), n = s_start_month - 1)
+    month_abbr_vector <- factor(forcats::fct_shift(f = lubridate::month(col_data, label = TRUE), n = s_start_month - 1),ordered = FALSE)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space before `ordered`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 2]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
     self$add_columns_to_data(col_name = col_name, col_data = dekad_val_vector)
   }
   if(dekad_abbr) {
-    month_abbr_vector <- forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
-	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 36
-	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 36, dekad_val_vector)
-	  dekad_abbr_vector <- paste(month_abbr_vector, dekad_val_vector, sep = "")
+    month_abbr_vector <- factor(forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1)),ordered = FALSE)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space before `ordered`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 3]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
     self$add_columns_to_data(col_name = col_name, col_data = dekad_val_vector)
   }
   if(dekad_abbr) {
-    month_abbr_vector <- forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
-	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 36
-	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 36, dekad_val_vector)
-	  dekad_abbr_vector <- paste(month_abbr_vector, dekad_val_vector, sep = "")
+    month_abbr_vector <- factor(forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1)),ordered = FALSE)
+	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 3
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space around `*`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 4]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
   if(dekad_abbr) {
-    month_abbr_vector <- forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
-	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 36
-	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 36, dekad_val_vector)
-	  dekad_abbr_vector <- paste(month_abbr_vector, dekad_val_vector, sep = "")
+    month_abbr_vector <- factor(forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1)),ordered = FALSE)
+	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 3
+	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 3, dekad_val_vector)
+	  month.list <- c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
+	  month_levels<-if (s_start_month == 1) month.list else c(tail(month.list, -s_start_month + 1), head(month.list, s_start_month - 1))
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space around `<-`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 5]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
+	  dekad_levels<-paste0(rep(month_levels, each = 3), 1:3)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space around `<-`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 6]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
+	  dekad_abbr_vector <- factor(paste(month_abbr_vector, dekad_val_vector, sep = ""),levels = dekad_levels)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space before `levels`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 7]
```diff
@@ -2250,10 +2253,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
   if(pentad_abbr) {
-	  month_abbr_vector <- forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
-	  pentad_val_vector <- ((as.integer(pentad(col_data))) - (s_start_month - 1)*6) %% 72
-	  pentad_val_vector <- ifelse(pentad_val_vector == 0, 72, pentad_val_vector)
-	  pentad_abbr_vector <- paste(month_abbr_vector, pentad_val_vector, sep = "")
+	  month_abbr_vector <-forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
+	  pentad_val_vector <- ((as.integer(pentad(col_data))) - (s_start_month - 1)*6) %% 6
+	  pentad_val_vector <- ifelse(pentad_val_vector == 0, 6, pentad_val_vector)
+	  month.list <- c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
+	  month_levels<-if (s_start_month == 1) month.list else c(tail(month.list, -s_start_month + 1), head(month.list, s_start_month - 1))
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space around `<-`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 8]
```diff
@@ -2250,10 +2253,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
+	  pentad_levels<-paste0(rep(month_levels, each = 6), 1:6)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space around `<-`

File: instat/static/InstatObject/R/data_object_R6.R [hunk 9]
```diff
@@ -2250,10 +2253,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
+	  pentad_abbr_vector <- factor(paste(month_abbr_vector, pentad_val_vector, sep = ""),levels = pentad_levels)
```
Comment (dannyparsons) [inline instat/static/InstatObject/R/data_object_R6.R]: space before `levels`

**Comment by AlexSananka** [comment]
@dannyparsons this is ready.

**Comment by maxwellfundi** [comment]
@dannyparsons  Formatting issues here seems to have been resolved.

File: instat/static/InstatObject/R/data_object_R6.R [hunk 10]
```diff
@@ -2236,10 +2236,13 @@ DataSheet$set("public","split_date", function(col_name = "", year_val = FALSE, y
...
   }
   if(dekad_abbr) {
-    month_abbr_vector <- forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1))
-	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 36
-	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 36, dekad_val_vector)
-	  dekad_abbr_vector <- paste(month_abbr_vector, dekad_val_vector, sep = "")
+    month_abbr_vector <- factor(forcats::fct_shift(f = (lubridate::month(col_data, label = TRUE)), n = (s_start_month - 1)), ordered = FALSE)
+	  dekad_val_vector <- ((as.numeric(dekade(col_data))) - (s_start_month - 1)*3) %% 3
+	  dekad_val_vector <- ifelse(dekad_val_vector == 0, 3, dekad_val_vector)
+	  month.list <- c("Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
```
Comment (shadrackkibet) [inline instat/static/InstatObject/R/data_object_R6.R]: Should this be `month_list`?

Comment (dannyparsons) [reply instat/static/InstatObject/R/data_object_R6.R]: yes it should to use the right naming convention.

**Pull request merged**